File: services/notification.ts
import axios from "axios";
import Cookies from "js-cookie";
import { CustomLogger } from "@/utils/logger";
import GatewayClient from "./gateway";

class NotificationClient {
  private baseUrl: string;
  private logger;
  private cookies: any;
  private gatewayClient: GatewayClient;
  public source?: string;

  constructor(source?: string) {
    this.baseUrl = process.env.BASE_URL || "";
    this.logger = new CustomLogger();
    this.gatewayClient = new GatewayClient();
    this.source = source;

    if (typeof window === "undefined") {
      // This code will only be executed on the server
      this.cookies = require("next/headers").cookies;
    } else {
      // This code will only be executed on the client
      this.cookies = Cookies;
    }
  }

  serverLogger(level: "info" | "debug" | "warn" | "error", message: string) {
    if (this.source === "server") {
      switch (level) {
        case "info":
          this.logger.info(message);
          break;
        case "debug":
          this.logger.debug(message);
          break;
        case "warn":
          this.logger.warn(message);
          break;
        case "error":
          this.logger.error(message);
          break;
        default:
          break;
      }
    }
  }

  async addUserNotification(payload: any) {
    try {
      const notificationUrl = `${this.baseUrl}/trade-notification/api/notifications`;
      const token = Cookies.get("token")
        ? Cookies.get("token")
        : this.cookies().get("token")?.value;
      if (!token) {
        throw Error("No token was provided. Failed to post notification");
      }

      const options = {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      };

      if (!payload.externalArgs) {
        payload.externalArgs = JSON.stringify({});
      }

      const response = await axios.post(notificationUrl, payload, options);
      const notification = response.data;

      return notification;
    } catch (error) {
      this.logger.error(`Failed to create notification : ${error}`);
    }
  }

  async addProductOwnerNotification(
    appSigninPayload: any,
    productOwner: string,
    data: any
  ) {
    try {
      const notificationUrl = `${this.baseUrl}/trade-notification/api/notifications`;

      // Get Product Owner User
      const signinPayload = appSigninPayload;

      const appToken = await this.gatewayClient.getToken(
        signinPayload.email,
        signinPayload.password
      );

      if (!appToken) {
        throw Error("No token was provided. Failed to post notification");
      }

      const productOwnerUser = await this.gatewayClient.getUserByEmail(
        productOwner,
        appToken
      );
      // const productOwnerImage = await getUserAvatar(productOwnerUser?.email as string ,appToken)

      // Post notification
      const options = {
        headers: {
          Authorization: `Bearer ${appToken}`,
        },
      };

      const payload = {
        userEmail: productOwnerUser?.email,
        username: productOwnerUser?.username,
        userImage: data.userImage,
        userId: productOwnerUser?.id,
        title: data.title,
        message: data.message,
        seen: false,
      };

      const response = await axios.post(notificationUrl, payload, options);
      const notification = response.data;

      return notification;
    } catch (error) {
      this.logger.error(`Failed to create notification : ${error}`);
    }
  }

  async getUserNotifications(email: string, userToken = null) {
    try {
      const notificationUrl = `${this.baseUrl}/trade-notification/api/notifications?userEmail=${email}`;
      const token = Cookies.get("token") ? Cookies.get("token") : userToken;
      if (!token) {
        throw Error("No token was provided. Failed to post notification");
      }

      const options = {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      };
      const response = await axios.get(notificationUrl, options);
      const notification = response.data;

      return notification;
    } catch (error) {
      this.logger.error(`Failed to get notifications : ${error}`);
    }
  }

  async updateNotification(notificationId: number, payload: any) {
    try {
      const notificationUrl = `${this.baseUrl}/trade-notification/api/notifications/${notificationId}`;
      const token = Cookies.get("token")
        ? Cookies.get("token")
        : this.cookies().get("token")?.value;
      if (!token) {
        throw Error("No token was provided. Failed to post notification");
      }

      const options = {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      };
      const response = await axios.patch(notificationUrl, payload, options);
      const notification = response.data;

      return notification;
    } catch (error) {
      this.logger.error(`Failed to update notification : ${error}`);
    }
  }

  async updateBulkNotification(payload: any) {
    try {
      const notificationUrl = `${this.baseUrl}/trade-notification/api/bulk_notifications`;
      const token = Cookies.get("token")
        ? Cookies.get("token")
        : this.cookies().get("token")?.value;
      if (!token) {
        throw Error("No token was provided. Failed to post notification");
      }

      const options = {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      };
      const response = await axios.patch(notificationUrl, payload, options);
      const notifications = response.data;

      return notifications;
    } catch (error) {
      this.logger.error(`Failed to update notifications : ${error}`);
    }
  }

  async deleteNotification(notificationId: number) {
    try {
      const notificationUrl = `${this.baseUrl}/trade-notification/api/notifications/${notificationId}`;
      const token = Cookies.get("token")
        ? Cookies.get("token")
        : this.cookies().get("token")?.value;
      if (!token) {
        throw Error("No token was provided. Failed to post notification");
      }

      const options = {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      };
      const response = await axios.delete(notificationUrl, options);
      const notification = response.data;

      return notification;
    } catch (error) {
      this.logger.error(`Failed to delete notification : ${error}`);
    }
  }
}

export default NotificationClient;

File: connectors/task_scheduler/notification.py
import json
import requests
from task_scheduler_app import settings
from task_scheduler_app.clients.gateway import Gateway
from task_scheduler_app.tools.helpers import *


class Notification:
    def __init__(self) -> None:
        self.notification_base_url = settings.NOTIFICATION_BASE_URL
        self.notification_url = f'{self.notification_base_url}/api/notifications'
        self.bulk_notification_url = f'{self.notification_base_url}/api/bulk_notifications'
        self.token = None

    def add_user_notification(self, payload):
        try:
            if not self.token:
                raise ValueError("No token was provided. Failed to post notification")

            headers = {}
            headers['Content-Type'] = 'application/json'
            headers['Authorization'] = f'Bearer {self.token}'

            if not payload.get("externalArgs", None):
                payload["externalArgs"] = json.dumps({})

            response = requests.post(url=self.notification_url, verify=False, json=payload, headers=headers)
            if response.status_code == 200:
                notification = response.json()
                return notification
            else:
                logger.error("=====> add notification failed --> failed")
                logger.error(response.status_code)
                logger.error(response.text)
            return None
        except Exception as error:
            logger.error(f"Failed to create notification: {error}")

    def add_target_user_notification(self, app_signin_payload, target_user, data):
        try:
            gateway = Gateway()
            target_user_user = gateway.get_user_by_email(email=target_user, use_admin_token=True, admin_credentials=app_signin_payload)
            if not gateway.token:
                raise ValueError("No token was provided. Failed to post notification")

            headers = {}
            headers['Content-Type'] = 'application/json'
            headers['Authorization'] = f'Bearer {gateway.token}'

            payload = {
                "userEmail": target_user_user["email"],
                "username": target_user_user["username"],
                "userImage": data["userImage"],
                "userId": target_user_user["id"],
                "title": data["title"],
                "message": data["message"],
                "seen": False,
            }

            response = requests.post(url=self.notification_url, verify=False, json=payload, headers=headers)
            if response.status_code == 201:
                notification = response.json()
                return notification
            else:
                logger.error("=====> add notification failed --> failed")
                logger.error(response.status_code)
                logger.error(response.text)
            return None

        except Exception as error:
            logger.error(f"Failed to create notification: {error}")

    def get_user_notifications(self,email):
        try:
            notification_url = f"{self.notification_base_url}/api/notifications/?userEmail={email}"
            if not self.token:
                raise ValueError("No token was provided. Failed to post notification")

            headers = {}
            headers['Content-Type'] = 'application/json'
            headers['Authorization'] = f'Bearer {self.token}'

            response = requests.get(url=notification_url, verify=False, headers=headers)
            if response.status_code == 200:
                notification = response.json()
                return notification
            else:
                logger.error("=====> get user notifications failed --> failed")
                logger.error(response.status_code)
                logger.error(response.text)
            return None

        except Exception as error:
            logger.error(f"Failed to get notifications: {error}")

    def update_notification(self, notification_id, payload):
        try:
            if not self.token:
                raise ValueError("No token was provided. Failed to post notification")

            headers = {}
            headers['Content-Type'] = 'application/json'
            headers['Authorization'] = f'Bearer {self.token}'

            response = requests.patch(url=f"{self.notification_url}/{notification_id}", verify=False, json=payload, headers=headers)
            if response.status_code == 200:
                notification = response.json()
                return notification
            else:
                logger.error("=====> update notification failed --> failed")
                logger.error(response.status_code)
                logger.error(response.text)
            return None

        except Exception as error:
            logger.error(f"Failed to update notification: {error}")

    def update_bulk_notification(self, payload):
        try:
            if not self.token:
                raise ValueError("No token was provided. Failed to post notification")

            headers = {}
            headers['Content-Type'] = 'application/json'
            headers['Authorization'] = f'Bearer {self.token}'

            response = requests.patch(url=self.bulk_notification_url, verify=False, json=payload, headers=headers)
            if response.status_code == 200:
                notifications = response.json()
                return notifications
            else:
                logger.error("=====> update notifications failed --> failed")
                logger.error(response.status_code)
                logger.error(response.text)
            return None

        except Exception as error:
            logger.error(f"Failed to update notifications: {error}")

    def delete_notification(self, notification_id):
        try:
            if not self.token:
                raise ValueError("No token was provided. Failed to post notification")

            headers = {}
            headers['Content-Type'] = 'application/json'
            headers['Authorization'] = f'Bearer {self.token}'

            response = requests.delete(url=f"{self.notification_url}/{notification_id}", verify=False, headers=headers)
            if response.status_code == 204:
                notification = response.json()
                return notification
            else:
                logger.error("=====> dlete notification failed --> failed")
                logger.error(response.status_code)
                logger.error(response.text)
            return None

        except Exception as error:
            logger.error(f"Failed to delete notification: {error}")